walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
 
 package com.google.sps;
 
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
+import com.google.sps.MeetingRequest;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, 
+                                     MeetingRequest request) {
+    // Get unavailable time for all required attendees in meeting request
+    ArrayList<TimeRange> unavailable = new ArrayList<>();
+    for (String attendee : request.getAttendees()) {
+      ArrayList<TimeRange> meetingTimes = getMeetingTimes(events, attendee);
+      unavailable = merge(unavailable, meetingTimes);
+    }
+
+    // Get time ranges that are longer than the requested meeting duration
+    Collection<TimeRange> result = getSortedAvailability(unavailable);
+    result.removeIf( 
+      timeRange -> timeRange.duration() < request.getDuration());
+
+    return result;
+  }
+
+  /**
+   * Adds the unavailable time ranges of a new attendee to the cumulative
+   * unavailable time ranges.
+   * Resolves conflicts by merging overlapping ranges into new ranges
+   */
+  private ArrayList<TimeRange> merge(ArrayList<TimeRange> result, 
+                                     ArrayList<TimeRange> other) {
+    if (result.isEmpty()) {
+      return other;
+    }
+    if (other.isEmpty()) {
+      return result;
+    }
+
+    result.addAll(other);
+    Collections.sort(result, TimeRange.ORDER_BY_START);
+
+    // Merge overlapping ranges
+    for (int i = 0; i < result.size()-1; i++) {
+      TimeRange current = result.get(i);
+      TimeRange next = result.get(i+1);
+
+      if (current.overlaps(next)) {
+        if (current.contains(next)) {
+          // Case 1:  |_______|
+          //            |__|
+          result.remove(next);
+        } else {
+          // Case 2:  |______|
+          //               |______|
+          result.set(i, TimeRange.fromStartEnd(current.start(), 
+                                               next.end(), false));
+          result.remove(next);
+          i--;
```

Comment:
You don't need to decrease i here.

Reply:
I think since I modify the object at i (I merged TimeRange at i and i+1 and put the new range at i), I need to consider that new range as well.
For example, scenario like
<img width="185" alt="Screen Shot 2020-07-01 at 5 27 23 PM" src="https://user-images.githubusercontent.com/37273127/86227795-29587780-bbc0-11ea-843c-d09e4058f365.png">
I need to merge all three together.

Reply:
Yes. you're right.